File: practices/listsumandmax/max.py
def is_empty(lst):
    return len(lst) == 0

def is_last(lst, index):
    return index == len(lst) - 1

def recursive_max(lst, index=0):
    if is_empty(lst):
        return 0  # or some base case; list won't be empty per problem
    if is_last(lst, index):
        return lst[index]

    current = lst[index]
    next_max = recursive_max(lst, index + 1)

    if current > next_max:
        return current
    else:
        return next_max

# ---- main ----
n = int(input())
numbers = [int(input()) for _ in range(n)]

print(recursive_max(numbers))

File: practices/listsumandmax/max.cpp
#include <bits/stdc++.h>
using namespace std;

bool list_empty(const vector<int>& lst) {
    return lst.empty();
}

bool list_last(const vector<int>& lst, int index) {
    return index == (int)lst.size() - 1;
}

int recursive_max(const vector<int>& lst, int index = 0) {
    if (list_empty(lst))
        return 0; // base case (though list will not be empty)
    if (list_last(lst, index))
        return lst[index];

    int next_max = recursive_max(lst, index + 1);
    if (lst[index] > next_max)
        return lst[index];
    else
        return next_max;
}

int main() {
    int N;
    cin >> N;
    vector<int> arr(N);
    for (int i = 0; i < N; i++)
        cin >> arr[i];

    cout << recursive_max(arr) << "\n";
    return 0;
}
